lib/internal/test_runner/test.js

```diff
@@ -413,7 +413,9 @@ class Test extends AsyncResource {
 
       this.pass();
     } catch (err) {
-      if (err?.code === 'ERR_TEST_FAILURE' && kIsNodeError in err) {
+      if (err?.code === 'ERR_TEST_FAILURE' && kIsNodeError in err && err.failureType === kTestTimeoutFailure) {
+        this.cancel(err);
+      } else if (err?.code === 'ERR_TEST_FAILURE' && kIsNodeError in err) {
         this.fail(err);
```

Comment:
To avoid duplication, can you check `err?.code === 'ERR_TEST_FAILURE' && kIsNodeError in err` once and then inside of that branch either call `cancel()` or `fail()` based on `err.failureType === kTestTimeoutFailure`.